### DISCUSSION TITLE
GSUB lookup flags

### DISCUSSION THREAD

**Discussion by seehuhn** [harfbuzz/harfbuzz#3538] [Q&A] [answered]
Hello,

I am trying to understand how lookup flags work in harfbuzz.
Here is an experiment I tried:

* I made an OpenType font which has the letters A-Z at glyph IDs 1, 2, ..., 26.
* For this test, I marked glyphs 2 ("B") and 4 ("D") as ligatures in the GDEF table.
  Otherwise the gdef table is empty.
* I apply a GSUB 5.1 lookup (Context Substitution Format 1) with the
  IGNORE_LIGATURES flat set, which matches for the input "ACE".
  Since B and D are classed as ligatures here, this matches at position 1
  of the string "ABCDE".
* The GSUB 5.1 lookup contains two Sequence lookup records:
  - The first sequence lookup record uses a GSUB 4.1 subtable
    at sequenceIndex 1 to replace "CD" -> "I".
    This applies successfully (but see question 1, below).
  - The second sequence lookup record tries to use a GSUB 1.1 subtable
    at sequenceIndex 2 to replace "E" -> "O".
    This does not work.

questions:

1. Why does the "CD" -> "I" substitution work?  Should this not be
   restricted to the input sequence "ACE", which has no "D"?
   The reason I was expecting this substitution to fail is that
   the spec at the very end of the section
   https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#lookup-table
   states that "the flags in the nested lookup can result in a **secondary level**
   of filtering".

2. Why does the second sequence lookup record, "E" -> "O", not apply?
   (It does apply if I change the first sequence lookup record to be of type 1, format 1.)

Details:

The font file (both as `.otf` and decoded via `ttx`) is here:

    https://m.seehuhn.de/hb9735/test9735.otf
    https://m.seehuhn.de/hb9735/test9735.ttx

I called harfbuzz via the hp-shape utility:

    > hb-shape test9735.otf ABCDE
    [gid1=0+1366|gid2=1+1366|gid9=2+817|gid5=4+1366]

Here gid9 is the "I" which was substituted and gid5 is the "E" which
was not replaced.

Many thanks,
Jochen

**Comment by behdad** [comment]
> 1. Why does the "CD" -> "I" substitution work?  Should this not be
>    restricted to the input sequence "ACE", which has no "D"?

No. The spec, clearly says (same paragraph you quoted):

"Lookup flags in the main lookup table are not considered while the nested lookups are being processed."

**Reply by seehuhn** [reply]
Ah, ok, this makes sense.

**Reply by seehuhn** [reply]
To double-check, I tried this out with a current version of Microsoft Word (Office 365 on Mac) and I can confirm that Word also has the behaviour you suggest.  Somehow they keep track of which glyph is which, and also manage to do the second substitution.

**Reply by behdad** [reply]
Thanks for checking.

**Reply by seehuhn** [reply]
Are you interested in cases where GSUB tables in Microsoft word behave differently from harfbuzz?  (For silly, constructed examples similar to the one above.)  If so, how should I report such cases?  Note them here?  File an issue?

**Reply by behdad** [reply]
Yes please. File an issue.

**Reply by seehuhn** [reply]
great, I've just filed #3545 .

**Reply by seehuhn** [reply]
... and #3556

**Reply by seehuhn** [reply]
The more I play with this, the more differences I find.  Here is another one.  My GSUB table does the following:

    GSUB_5: -ligs "AAA" -> 1@0
    GSUB_5: "AL" -> 2@1 3@1
    GSUB_4: "LA" -> "K"
    GSUB_1: "L" -> "B"

where `-ligs` means that lookup 0 has the `IGNORE_LIGATURES` flag set, and `1@0` means "apply lookup 1 and position 0 as a nested lookup" (similar for 2@1, ...).  In my font I have marked the `L` glyph as a ligature.  Then the input `ALAA` gives `AKA` in HarfBuzz and on MacOS, but gives `ABAA` on Windows.  Apparently Windows applies lookup 1 at position 0 as planned, but then somehow does not apply lookup 2 at position 1 (so that then lookup 3 matches, which is applied fine).  I'd be happy to file a proper issue and/or send the font file if this is of interest.

**Reply by behdad** [reply]
> I'd be happy to file a proper issue and/or send the font file if this is of interest.

Issue is good, even if we won't fix it. Thanks.

**Comment by behdad** [comment, marked as answer]
> 2\. Why does the second sequence lookup record, "E" -> "O", not apply?
>     (It does apply if I change the first sequence lookup record to be of type 1, format 1.)

That's a deficiency of HarfBuzz. In the meantime, here's the code for it:
https://github.com/harfbuzz/harfbuzz/blob/9e59c401e05018eacc09d277251f692a517652c8/src/hb-ot-layout-gsubgpos.hh#L1444-L1467

When the recursed lookup has a different LookupFlag is an example of when the removed items are not the same as matched-items, so that assumption fails. Is really hard to fix this unfortunately.

**Reply by seehuhn** [reply]
Thanks for the link to the code (I keep getting lost in the harfbuzz code base).  And I agree, if the nested lookups really operate on the original sequence (as opposed to the matched input), it is very hard to track which of the original input glyphs are still there after earlier lookups have modified the glyph sequence :(

**Comment by behdad** [comment]
One way to work around this for you is to change the order of the two substitutions.

**Comment by behdad** [comment]
https://github.com/harfbuzz/harfbuzz/commit/a85461b9b6367d5ce313c800f9efc6a1ad750616
